Duecredits implemented.  Only used two citations as of now. (1) Michaud Agrawal (2011) (2) R. Gowers (2016)

ayushsuhane opened the pull request:
Issue #412 : Initial commit for the citations when any script is executed as 'python -m duecredit test.py 


Changes made in this Pull Request: The changes are as follows:
(1) due.py in package/MDAnalysis/ which calls the duecredit  module
(2) Modification in universe.py, which imports due.py and attached decorators of citation to the __init__.py.
 -

orbeckst commented on package/MDAnalysis/core/universe.py:
```diff
@@ -105,6 +105,7 @@
                      AtomGroup, ResidueGroup, SegmentGroup)
 from .topology import Topology
 from .topologyattrs import AtomAttr, ResidueAttr, SegmentAttr
+from ..due import due,Doi,BibTeX
```
PEP8 style: space after comma

orbeckst commented on package/MDAnalysis/core/universe.py:
```diff
@@ -213,7 +214,19 @@ class Universe(object):
 
     """
-
+    @due.dcite(BibTeX(("""
```
The citations should go into the top level `MDAnalysis/__init__.py` – or what is your rationale for putting them here?

orbeckst replied:
Actually, ignore my comment. It makes sense to add the citations to `Universe`. 

However, is there a way to associate these citations globally with the library because someone might just use `MDAnalysis.lib.formats` but we would still want them to cite us.

orbeckst commented on package/MDAnalysis/core/universe.py:
```diff
@@ -213,7 +214,19 @@ class Universe(object):
+            }
+            """)), 
+            description="Gowers 2016", path="MDAnalysis/")
+    @due.dcite(Doi("10.1002/jcc.21787"),description="MDAnalysis : A Toolkit for the Analysis of Molecular Dynamics Simulations, J. Comput. Chem. 32 (2011), 2319-2327")
```
PEP 8: space after comma

orbeckst replied:
See also https://github.com/MDAnalysis/mdanalysis/wiki/Style-Guide#coding-style (eg length of lines)

ayushsuhane commented:
Moving the citation informations to __init__.py causes an error in testsuite/MDAnalysisTests/analysis/test_encore.py. I am unable to understand the linkages here. @orbeckst  Can you help me figure it out ? Thanks.

I am yet to write the test case for duecredit, once this issue is resolved.

ayushsuhane commented:
![image](https://user-images.githubusercontent.com/34154224/36938924-a9843822-1edd-11e8-9405-96e6317976a3.png)
When I run  the same test on my machine, it runs successfully.

ayushsuhane commented:
@orbeckst  Hi, I am having problem with writing a test case. Majorly its because, duecredit outputs  the citations on the terminal(as well as in a .p file) once the script is terminated. Do you have any suggestions, how can I cover this in a test case?

orbeckst commented:
The error in encore is spurious and not your fault. The results in this test are very sensitive and this test can fail randoml (it's not a great test and should be changed...).

orbeckst commented:
If `duecredit` is not installed then the warning is logged and it does not work (as far as I can tell). I really do not want this warning to show up for users that install MDA because it indicates that something is not working and most users get immediately concerned.

We should probably make duecredit a dependency for pip and conda (which will require additional packages downstream, namely citeproc and lxml as far as I can tell).

If creating additional dependencies is not a good solution then the warning should only be issued once one tries to get citations. 

We'll have to discuss what exactly we want to do.

orbeckst commented:
duecredit is easily installable with pip but there's no conda package.

Given that the [recommended way to use it](https://github.com/duecredit/duecredit#examples) is to install duecredit yourself and then run
```
python -m duecredit mdascript.py
```
This will then write a hidden `.duecredit.p` summary file that is further processed with the `duecredit` command line utility
```
duecredit summary --format=bibtex
```

Given the recommended workflow, we probably don't have to install duecredit together with MDAnalysis (although it would be convenient). 

I found that setting the environment variable `DUECREDIT_ENABLE=yes` will create the summary file if the duecredit package is installed whenever the code is run:
```
DUECREDIT_ENABLE=yes python mdascript.py
```

(Maybe this can be used in tests?)

orbeckst commented:
@ayushsuhane in reply to your [message on the mailing list](https://groups.google.com/d/msg/mdnalysis-devel/XYZB6gzMI2g/Uk3DMR-lAwAJ):

# Approaches to testing
After playing around with duecredit I understand the problem that you're raising: duecredit uses atexit to only write its output `.duecredit.p` at the end of a script. This makes it hard to test. I have a few suggestions:

## Approach 1: testing the collector
If we run the whole testsuite with environment variable `DUECREDIT_ENABLE=yes` (or do a `os.environ['DUECREDIT_ENABLE'] = 'yes'` *before* importing MDAnalysis for the first time in the tests) then we can test the collector itself:
```python
>>> MDAnalysis.due.active
True
>>> MDAnalysis.due.citations
{CitationKey(path='MDAnalysis/', entry_key='10.1002/jcc.21787'): Citation(Doi('10.1002/jcc.21787', key='10.1002/jcc.21787'), description='MDAnalysis : A Toolkit for the Analysisof Molecular Dynamics Simulations,J. Comput. Chem. 32 (2011), 2319-2327', path='MDAnalysis/', cite_module=True),
 CitationKey(path='MDAnalysis/', entry_key='gowers2016'): Citation(BibTeX('@inproceedings{gowers2016,title={MDAnalysis: A Python package for the rapid analysisof molecular dynamics simulations},author={R. J. Gowers and M. Linke andJ. Barnoud and T. J. E. Reddy and M. N. Meloand S. L. Seyler and D. L. Dotson and J. Domanski andS. Buchoux and I. M. Kenney and O. Beckstein},journal={Proceedings of the 15th Python in Science Conference},pages={102-109},year={2016},editor={In S. Benthall and S. Rostrup},note={Austin, TX, SciPy.}}', key='gowers2016'), description='Gowers 2016', path='MDAnalysis/', cite_module=True),
 CitationKey(path='numpy', entry_key='van2011numpy'): Citation(BibTeX('@article{van2011numpy,\n        title={The NumPy array: a structure for efficient numerical computation},\n        author={Van Der Walt, Stefan and Colbert, S Chris and Varoquaux, Gael},\n        journal={Computing in Science \\& Engineering},\n        volume={13},\n        number={2},\n        pages={22--30},\n        year={2011},\n        publisher={AIP Publishing}\n        }', key='van2011numpy'), description='Scientific tools library', path='numpy', cite_module=True),
 CitationKey(path='scipy', entry_key='JOP+01'): Citation(BibTeX('@Misc{JOP+01,\n      author =    {Eric Jones and Travis Oliphant and Pearu Peterson and others},\n      title =     {{SciPy}: Open source scientific tools for {Python}},\n      year =      {2001--},\n      url = "http://www.scipy.org/",\n      note = {[Online; accessed 2015-07-13]}\n    }', key='JOP+01'), description='Scientific tools library', path='scipy')}
>>> MDAnalysis.due.citations.keys()
dict_keys([CitationKey(path='numpy', entry_key='van2011numpy'), CitationKey(path='scipy', entry_key='JOP+01'), CitationKey(path='MDAnalysis/', entry_key='gowers2016'), CitationKey(path='MDAnalysis/', entry_key='10.1002/jcc.21787')])
```
Basically, I would check that the citations that *we* added are present. Once this is assured, I think we can  assume that processing the citation with duecredit will also work.

You can also run (in a tmpdir)
```python
>>> MDAnalysis.due.dump()


DueCredit Report:
- Gowers 2016, MDAnalysis : A Toolkit for the Analysisof Molecular Dynamics Simulations,J. Comput. Chem. 32 (2011), 2319-2327 / MDAnalysis/ (v None) [1, 2]
- Scientific tools library / numpy (v 1.14.1) [3]

2 packages cited
0 modules cited
0 functions cited

References
----------

[1] Gowers, R.J. et al., 2016. MDAnalysis: A Python package for the rapid analysisof molecular dynamics simulations. In I. S. Benthall & S. Rostrup, eds. Proceedings of the 15th Python in Science Conference. pp. 102–109.
[2] Michaud-Agrawal, N. et al., 2011. MDAnalysis: A toolkit for the analysis of molecular dynamics simulations. Journal of Computational Chemistry, 32(10), pp.2319–2327.
[3] Van Der Walt, S., Colbert, S.C. & Varoquaux, G., 2011. The NumPy array: a structure for efficient numerical computation. Computing in Science & Engineering, 13(2), pp.22–30.
```
to get the summary output. The output could be captured and one could check that the summary file has been generated.

## Approach 2: testing duecredit
We could run an external process in a tmpdir (the following code won't work because I don't remember the call signature for `popen`)
```python
buf = StringIO()
popen(['python',  '-m',  'duecredit', '-c', 'import MDAnalysis'], stdout=buf)
```
and then check that we have the proper `.duecredit.p` file and that the output is as expected. In this case using StringIO(), output capturing, and monkeypatching are all acceptable. 

# conclusions
I would rather just test the collector itself (approach 1). That should be sufficient and also allow for cleaner tests.

orbeckst commented on package/MDAnalysis/__init__.py:
```diff
@@ -200,3 +200,29 @@
+            "journal={Proceedings of the 15th Python in Science Conference},"
+            "pages={102-109},"
+            "year={2016},"
+            "editor={In S. Benthall and S. Rostrup},"
```
just
```
"editor={S. Benthall and S. Rostrup},"
```
(no "In"_

orbeckst commented on package/MDAnalysis/__init__.py:
```diff
@@ -200,3 +200,29 @@
+            "note={Austin, TX, SciPy.}"
+            "}"
+            )), 
+            description="Gowers 2016", path="MDAnalysis/", cite_module=True)
```
Please change description to

"Description of recent updates and new functionality in MDAnalysis since release 0.7.2 (March 2011) up to release 0.16.0 (April 2017). Please cite together with Michaud-Agrawal et al 2011."

orbeckst commented on package/MDAnalysis/__init__.py:
```diff
@@ -200,3 +200,29 @@
+            )), 
+            description="Gowers 2016", path="MDAnalysis/", cite_module=True)
+due.cite(Doi("10.1002/jcc.21787"),
+         description="MDAnalysis : A Toolkit for the Analysis"
```
Please change description to

"Description of MDAnalysis up to release 0.7.2 (March 2011), including the basic philosophy of the library and the LeafletFinder algorithm. Please cite together with Gowers et al 2016."

orbeckst commented on package/MDAnalysis/__init__.py:
```diff
@@ -200,3 +200,29 @@
+            "@inproceedings{gowers2016,"
+            "title={MDAnalysis: A Python package for the rapid analysis"
+            "of molecular dynamics simulations}," 
+            "author={R. J. Gowers and M. Linke and"
```
add space after "and": 

"and "

(otherwise the lines run into each other)

orbeckst commented on package/MDAnalysis/__init__.py:
```diff
@@ -200,3 +200,29 @@
+            "title={MDAnalysis: A Python package for the rapid analysis"
+            "of molecular dynamics simulations}," 
+            "author={R. J. Gowers and M. Linke and"
+            "J. Barnoud and T. J. E. Reddy and M. N. Melo"
```
add trailing space 

"Melo "

(otherwise the lines run into each other)

orbeckst commented on package/MDAnalysis/__init__.py:
```diff
@@ -200,3 +200,29 @@
+            "of molecular dynamics simulations}," 
+            "author={R. J. Gowers and M. Linke and"
+            "J. Barnoud and T. J. E. Reddy and M. N. Melo"
+            "and S. L. Seyler and D. L. Dotson and J. Domanski and"
```
add trailing space 

"and "

(otherwise the lines run into each other)

orbeckst commented on package/MDAnalysis/__init__.py:
```diff
@@ -200,3 +200,29 @@
+         "J. Comput. Chem. 32 (2011), 2319-2327",
+         path="MDAnalysis/",cite_module=True)
+
+
```
Remove `Doi` and `BibTeX` from namespace
```python
del Doi, BibTeX
```

orbeckst commented on package/MDAnalysis/core/universe.py:
```diff
@@ -106,6 +106,7 @@
 from .topology import Topology
 from .topologyattrs import AtomAttr, ResidueAttr, SegmentAttr
 
+
```
remove empty line change

orbeckst commented on package/MDAnalysis/due.py:
```diff
@@ -0,0 +1,73 @@
+except Exception as e:
+    if type(e).__name__ != 'ImportError':
+        import logging
+        logging.getLogger("duecredit").error(
```
This error message is not good. I tried (with `duecredit` not installed) under Python 3.6
```
DUECREDIT_ENABLE=yes python -c "import MDAnalysis"
```
and got
```
Failed to import duecredit due to No module named 'duecredit'
```
I *think* this error message should not have even appeared. This comes from a `ModuleNotFoundError` as far as I can tell ,which is a subclass of `ImportError`. 

How about doing something like

```python
except ImportError:
   pass
except Exception as err:
   import logging
   logging.getLogger("duecredit").error("Failed to import duecredit due to {}", str(e))
```

orbeckst replied:
Please also address this comment.

ayushsuhane replied:
I have addressed other issues, but I am not sure if I get this correctly.  If anything is ran with DUECREDIT_ENABLE = true, then shouldn't it be obvious that the user has duecredit installed. I think the maximum we can do is to show a warning that either duecredit needs to be installed or the change the environment variable.  What do you think? @orbeckst  . If it is not installed and DUECREDIT_ENABLE is not set, then it wouldn't give an error.

orbeckst replied:
The original intent of the code was, if I read it correctly, to *not show a warning when the module cannot be imported* (which is what duecredit promises). What I am saying is that this intended behavior is not working because the test `if type(e).__name__ != 'ImportError':` is wrong (at least in 3.6): it does not catch the `ModuleNotFoundError` whereas just catching `ImportError` will cleanly accomplish the goal.

By the way: when I import MDAnalysis with your current code and when `duecredit` is *not installed* I get
```python
>>> import MDAnalysis
Failed to import duecredit due to No module named 'duecredit'
>>> import os
>>> os.environ['DUECREDIT_ENABLE']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/oliver/anaconda3/envs/mda_develop/lib/python3.6/os.py", line 669, in __getitem__
    raise KeyError(key) from None
KeyError: 'DUECREDIT_ENABLE'
```
i.e., most users will see this warning. The behavior needs to be that users are not bothered with this import error.

orbeckst commented:
For the tests I would suggest to go with approach 1. We can set the environment variable in `.travis.yml` or even better in `MDAnalysisTests/__init__.py` *before* any `Import MDAnalysis` statements. 

We need to `pip install duecredit` within `.travis.yml` --- maybe you can find out yourself where to put this or ask @kain88-de. 

You can then write a test to check the duecredit collector directly, as outlined in my comment above.

ayushsuhane commented:
@orbeckst Hi, I made the changes, and the test cases runs successfully on my machine. However, I am not too sure about travis, whether it is a good way to install duecredit, but it works. @kain88-de  Can you check if this is alright?
@orbeckst I am not sure whether the tests are up and running on travis. I am not sure where to check for the test output. I added a skip decorator in pytest if duecredit is not installed. Should it be done any other way?

orbeckst commented on testsuite/MDAnalysisTests/utils/test_duecredits.py:
```diff
@@ -0,0 +1,60 @@
+import logging
+import pytest
+import os,sys
+os.environ['DUECREDIT_ENABLE']='yes'
```
I don't think that setting it here will make a difference because MDAnalysis has already been imported earlier. You can remove this one.

Setting it in travis.yml is safest but maybe it also works in `MDAnalysisTests/__init__.py` because MDAnalysis is not imported earlier than that.

orbeckst commented on testsuite/MDAnalysisTests/utils/test_duecredits.py:
```diff
@@ -0,0 +1,60 @@
+import os,sys
+os.environ['DUECREDIT_ENABLE']='yes'
+
+logger = logging.getLogger("MDAnalysisTests.utils")
```
Don't use the logger here, we don't use it for tests.

orbeckst commented on testsuite/MDAnalysisTests/utils/test_duecredits.py:
```diff
@@ -0,0 +1,60 @@
+
+logger = logging.getLogger("MDAnalysisTests.utils")
+#Testing the import#
+try:
```
Don't protect `import MDAnalysis.due` – if the test is present then `MDAnalysis.due` is present. Otherwise this should just fail horribly and that would be ok.

orbeckst commented on testsuite/MDAnalysisTests/utils/test_duecredits.py:
```diff
@@ -0,0 +1,60 @@
+    logger.info('Could not find due.py,all tests will be skipped')
+    
+
+try:
```
Don't protect `duecredit`.

Instead use [pytest.importorskip](https://docs.pytest.org/en/latest/skipping.html#skipping-on-a-missing-import-dependency) below.

orbeckst commented on testsuite/MDAnalysisTests/utils/test_duecredits.py:
```diff
@@ -0,0 +1,60 @@
+    os.environ['DUECREDIT_ENABLE']='no'
+
+class TestDuecredits():
+    @pytest.mark.skipif(os.environ['DUECREDIT_ENABLE']=='no',
```
You might be able to put the decorators on the class and then you don't have to repeat them.

orbeckst replied:
We want to skip if we cannot import `duecredit`, e.g. with [pytest.importorskip](https://docs.pytest.org/en/latest/skipping.html#skipping-on-a-missing-import-dependency), or if it has been explicitly disabled.
```python
@pytest.importorskip("duecredit")
@pytest.skipif(os.environ.get('DUECREDIT_ENABLE', "yes").lower() in ("no", "0", "false"), reason="duecredit is explicitly disabled with DUECREDIT_ENABLE=no")
```
(I think the above values for DUECREDIT_ENABLE are all properly recognized but check the duecredit docs/sources)

orbeckst commented on testsuite/MDAnalysisTests/utils/test_duecredits.py:
```diff
@@ -0,0 +1,60 @@
+    @pytest.mark.skipif(os.environ['DUECREDIT_ENABLE']=='no',
+                        reason="Duecredit is not installed, Skipping")    
+    def test_duecredit_active(self):
+        import MDAnalysis as mda
```
import at the top level of the test; it will not run duecredit because MDAnalysis has already been imported and all that Python is doing is aliasing the module into this name space.

orbeckst commented on testsuite/MDAnalysisTests/utils/test_duecredits.py:
```diff
@@ -0,0 +1,60 @@
+    @pytest.mark.skipif(os.environ['DUECREDIT_ENABLE']=='no',
+                        reason="Duecredit is not installed,Skipping")
+    def test_duecredit_collector_citations(self):
+        import MDAnalysis as mda
```
import at the top level of the test; it will not run duecredit because MDAnalysis has already been imported and all that Python is doing is aliasing the module into this name space.

orbeckst commented on .travis.yml:
```diff
@@ -33,6 +33,7 @@ env:
     - CONDA_CHANNEL_PRIORITY=True
     - NUMPY_VERSION=stable
     - INSTALL_HOLE="true"
+    - DUECREDIT_ENABLE=yes
```
Try it with global env var here.

Then try it with setting
```python
os.environ["DUECREDIT_ENABLE"] = "yes"
```
in [MDAnalysisTests/__init__.py](https://github.com/MDAnalysis/mdanalysis/blob/develop/testsuite/MDAnalysisTests/__init__.py) (and *not* in travis.yml).

orbeckst commented:
The `due.py` file is only covered to ~44%: https://coveralls.io/builds/15900857/source?filename=%2Fhome%2Ftravis%2Fminiconda%2Fenvs%2Ftest%2Flib%2Fpython3.5%2Fsite-packages%2FMDAnalysis%2Fdue.py

Given that this is a stub that we ddidn't write or modified we can exclude it from coverage by adding a line to `.coveragerc` 
```
omit =
        ...
        due.py
```

orbeckst commented:
Actually, it would be better to also have a test *without duecredit installed* to test that, well, nothing happens. That would probably test more of `due.py`. Do you have an idea how we can do this? Maybe mock-out the installed `duecredit` and re-initialized the collector??

I don't want to run a whole test suite just without duecredit installed because (1) wastes time and (2) does not accumulate coverage.

orbeckst commented on package/MDAnalysis/__init__.py:
```diff
@@ -200,3 +200,37 @@
+            "}"
+            )), 
+            description=
+         "Description of recent updates and new functionality"
```
add space at end of string

orbeckst commented on package/MDAnalysis/__init__.py:
```diff
@@ -200,3 +200,37 @@
+            )), 
+            description=
+         "Description of recent updates and new functionality"
+         "in MDAnalysis since release 0.7.2 (March 2011) up to"
```
add space at end of string

orbeckst commented on testsuite/MDAnalysisTests/utils/test_duecredits.py:
```diff
@@ -0,0 +1,44 @@
+from __future__ import absolute_import
+import logging
+import pytest
+import os
```
very minor style thing (PEP 8): re-order imports as
```python
from __future__ ...

import os
import pytest

import MDAnalysis as mda
```

Remove `import logging` (not needed)

orbeckst commented on .travis.yml:
```diff
@@ -95,6 +96,8 @@ install:
   - git clone git://github.com/astropy/ci-helpers.git
   - source ci-helpers/travis/setup_conda.sh
+  - pip install --upgrade pip
+  - pip install -v duecredit
```
Please remove `-v` to reduce clutter.

ayushsuhane commented:
I modified the files according to the review. Sorry for so many commits, I understand it looks ugly and maybe makes it more difficult for you to review.  
Second last commit is all the changes for duecredit, however it was showing a conflict in AUTHORS and CHANGELOG since I was not using updated files. I tried modifying them, but in the process all the files modified from the time I forked this repository, were also modified.   Please notify me if I need to do anything else in this PR. Once this is closed, I can work on adding additional references on modules as well as a test on duecredit itself.

orbeckst commented on testsuite/MDAnalysisTests/__init__.py:
```diff
@@ -101,7 +101,7 @@
 import os
 import sys
-
+os.environ['DUECREDIT_ENABLE']='yes'
```
Please add a comment that this is for testing duecredit in utils/test_duecredits.py

orbeckst commented on testsuite/MDAnalysisTests/utils/test_duecredits.py:
```diff
@@ -0,0 +1,44 @@
+import pytest
+
+import MDAnalysis as mda
+
```
Please add a comment saying that you're setting `DUECREDIT_ENABLE=yes` in `__init__.py`, just so that it is clear later how these things are connected.

orbeckst commented on package/doc/sphinx/source/documentation_pages/references.rst:
```diff
@@ -111,6 +111,29 @@ If you use the streamline visualization in
 
 .. _`10.1039/c3fd00145h`: https://doi.org/10.1039/c3fd00145h
 
+Citations using Duecredit
```
Well written, just small formatting changes below.

orbeckst commented on package/doc/sphinx/source/documentation_pages/references.rst:
```diff
@@ -111,6 +111,29 @@ If you use the streamline visualization in
+=========================
+
+Citations can be automatically generated using Duecredit, depending on the
+packages used. Duecredit is easy to install via pip. Simply type:
```
In restructured text, make this a code block:
```reST
... Simply type:

.. code-block:: bash

   pip install duecredit
```

orbeckst commented on package/doc/sphinx/source/documentation_pages/references.rst:
```diff
@@ -111,6 +111,29 @@ If you use the streamline visualization in
+
+duecredit will remain an optional dependency, i.e. your software will work
+correctly even without duecredit installed. List of citations for 
+yourscript.py can be obtained using simple commands.
```
Use code block:

```rest

... using simple commands
.. code-block:: bash

   cd /path/to/yourmodule
   python -m duecredit yourscript.py
```

orbeckst commented on package/doc/sphinx/source/documentation_pages/references.rst:
```diff
@@ -111,6 +111,29 @@ If you use the streamline visualization in
+python -m duecredit yourscript.py
+
+**Or**
+
```
```rest
.. code-block:: bash

   DUECREDIT-ENABLE=yes python yourscript.py
```

orbeckst commented on package/doc/sphinx/source/documentation_pages/references.rst:
```diff
@@ -111,6 +111,29 @@ If you use the streamline visualization in
+
+You can also display them in BibTeX format, using:
+
+duecredit summary --format=bibtex 
```
```rest
.. code-block:: bash

   duecredit summary --format=bibtex
```

orbeckst commented:
@ayushsuhane nice work, thank you for your first contribution!

I will fix it up so that it merges properly.

orbeckst commented:
@ayushsuhane I cleaned up the history but I cannot push to your branch (or force-push) so I created  PR #1822. It will count as *your* PR that you will have merged. I am using #1822 to finalize.

Thank you for your first contribution!

ayushsuhane commented:
@orbeckst  Did I disable anything on my branch? Or is it normal to not to be able to push into another user's branch. 

Thanks for all the clear reviews. I learnt a lot. Looking forward to work with you on another PR.

orbeckst commented:
>  Did I disable anything on my branch? Or is it normal to not to be able to push into another user's branch.

I don't know, unless you uncheck a box, pushing to PR branches works on GitHub. However, I tried a force-push with history rewriting so this might have been blocked (which is even kind of sensible). I din't have time to follow up in depth so I just quickly made the new PR.

Thanks for having been so responsive to all comments.
